Mathlib/Tactic/Linter/FlexibleLinter.lean

```diff
@@ -0,0 +1,384 @@
+* if the unique name of the `FVar` is preserved, then we use that;
+* otherwise, if the `userName` of the `FVar` is preserved, then we use that;
+* if neither is preserved, we drop the ball and stop tracking the `FVarId`.
+
+For the `MVar`s, we use the information of `Lean.Elab.TacticInfo.goalsBefore` and
+`Lean.Elab.TacticInfo.goalsAfter`.
+By looking at the `mvar`s that are either only "before" or only "after", we focus on the
+"active" goals.
+We then propagate all the `FVarId`s that were present in the "before" goals to the "after" goals,
+while leaving untouched the ones in the "inert" goals.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Linter
+
+/-- The flexible linter makes sure that "rigid" tactics do not follow "flexible" tactics. -/
+register_option linter.flexible : Bool := {
+  defValue := false
+  descr := "enable the flexible linter"
+}
+
+namespace flexible
+
+/-- `flexible? stx` is `true` on syntax that takes a "wide" variety of inputs and modifies
+them in possibly unpredictable ways.
+
+The prototypical flexible tactic is `simp`
+The prototypical non-flexible tactic `rw`.
+`simp only` is also non-flexible. -/
+def flexible? : Syntax → Bool
+  | .node _ ``Lean.Parser.Tactic.simp #[_, _, _, only?, _, _] => only?[0].getAtomVal != "only"
+  | .node _ ``Lean.Parser.Tactic.simpAll #[_, _, _, only?, _] => only?[0].getAtomVal != "only"
+  | _ => false
+
+end flexible
+
+end Mathlib.Linter
+
+section goals_heuristic
+namespace Lean.Elab.TacticInfo
+
+/-!
+###  Heuristics for determining active and inactive goals
+
+The two definitions `activeGoalsBefore`, `activeGoalsAfter` extract a list of
+`MVarId`s attempting to determine which on which goals the tactic `t` is acting.
+This is mostly based on the heuristic that the tactic with "change" an `MVarId`.
+-/
+
+/-- `activeGoalsBefore t` are the `MVarId`s before the `TacticInfo` `t` that "disappear" after it.
+They should correspond to the goals in which the tactic `t` performs some action. -/
+def activeGoalsBefore (t : TacticInfo) : List MVarId :=
+  t.goalsBefore.filter (·.name ∉ t.goalsAfter.map (·.name))
+
+/-- `activeGoalsAfter t` are the `MVarId`s after the `TacticInfo` `t` that were not present before.
+They should correspond to the goals changed by the tactic `t`. -/
+def activeGoalsAfter (t : TacticInfo) : List MVarId :=
+  t.goalsAfter.filter (·.name ∉ t.goalsBefore.map (·.name))
+
+end Lean.Elab.TacticInfo
+end goals_heuristic
+
+namespace Mathlib.Linter.flexible
+
+variable (take? : Syntax → Bool) in
+-- also returns the preceding goals that change.  is there only one always?
+/-- `extractCtxAndGoals take? tree` takes as input a function `take? : Syntax → Bool` and
+an `InfoTree` and returns the array of pairs `(stx, mvars)`,
+where `stx` is a syntax node such that `take? stx` is `true` and
+`mvars` are the goals that have been "created" by the tactic in `stx`.
+
+A typical usage is to find the goals following a `simp` application. -/
```

Comment:
Sounds really nice and clear... but I only see one reference to it below, where `take` is always true. :thinking: